controllers/receiver_controller.go

```diff
@@ -55,53 +62,116 @@ type ReceiverReconciler struct {
-	if receiver.Status.URL == receiverURL && isReady && receiver.Status.ObservedGeneration == receiver.Generation {
-		return ctrl.Result{}, nil
-	}
+		// Determine if the resource is still being reconciled, or if it has stalled, and record this observation
+		if retErr == nil && (result.IsZero() || !result.Requeue) {
+			// We are no longer reconciling
+			conditions.Delete(receiver, meta.ReconcilingCondition)
+
+			// We have now observed this generation
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+
+			readyCondition := conditions.Get(receiver, meta.ReadyCondition)
+			switch readyCondition.Status {
+			case metav1.ConditionFalse:
+				// As we are no longer reconciling and the end-state is not ready, the reconciliation has stalled
+				conditions.MarkStalled(receiver, readyCondition.Reason, readyCondition.Message)
+			case metav1.ConditionTrue:
+				// As we are no longer reconciling and the end-state is ready, the reconciliation is no longer stalled
+				conditions.Delete(receiver, meta.StalledCondition)
+			}
+		}
 
-	receiver = v1beta1.ReceiverReady(receiver,
-		v1beta1.InitializedReason,
-		"Receiver initialised with URL: "+receiverURL,
-		receiverURL)
-	if err := r.patchStatus(ctx, req, receiver.Status); err != nil {
-		return ctrl.Result{Requeue: true}, err
-	}
+		// Finally, patch the resource
+		if err := patchHelper.Patch(ctx, receiver, patchOpts...); err != nil {
+			retErr = errors.NewAggregate([]error{retErr, err})
+		}
 
-	log.Info("Receiver initialised")
+		// Always record readiness and duration metrics
+		r.Metrics.RecordReadiness(ctx, receiver)
+		r.Metrics.RecordDuration(ctx, receiver, start)
 
-	return ctrl.Result{}, nil
+	}()
+
+	return r.reconcile(ctx, receiver)
 }
 
 func (r *ReceiverReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return r.SetupWithManagerAndOptions(mgr, ReceiverReconcilerOptions{})
+}
+
+func (r *ReceiverReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, opts ReceiverReconcilerOptions) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1beta1.Receiver{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: opts.MaxConcurrentReconciles}).
 		Complete(r)
 }
 
+/// reconcile steps through the actual reconciliation tasks for the object, it returns early on the first step that
+// produces an error.
+func (r *ReceiverReconciler) reconcile(ctx context.Context, obj *v1beta1.Receiver) (ctrl.Result, error) {
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(obj, meta.ProgressingReason, "")
+
+	token, err := r.token(ctx, obj)
+	if err != nil {
+		conditions.MarkFalse(obj, meta.ReadyCondition, v1beta1.TokenNotFoundReason, err.Error())
+		return ctrl.Result{}, err
+	}
+
+	receiverURL := fmt.Sprintf("/hook/%s", sha256sum(token+obj.Name+obj.Namespace))
+
+	// Nothing has changed so return early
+	if obj.Status.URL == receiverURL && obj.Status.ObservedGeneration == obj.Generation {
+		return ctrl.Result{}, nil
+	}
```

Comment:
Is this check idempotent? E.g. if L150 fails due to a transient error and records `Ready == False`, wouldn't this prevent it from recovering to `Ready==True`?

Reply:
Good catch. It does prevent it from recovering. I will remove the check.